pkg/rpc/server/server.go

```diff
@@ -119,6 +143,14 @@ func New(chain core.Blockchainer, conf rpc.Config, coreServer *network.Server, l
 		coreServer: coreServer,
 		log:        log,
 		https:      tlsServer,
+		shutdown:   make(chan struct{}),
+
+		subscribers: make(map[*subscriber]bool),
+		// These are NOT buffered to preserve original order of events.
```

Comment:
I dislike this very much, because the order is implicit and depends on the order of sends in `core` (?) and/or timing assumptions (?). It is ok for now if it works, but in future we need something less obscure. Maybe a single channel with a mask containing event types we are interested in.
And it surely needs more explanation.

Reply:
> the order is implicit

Not exactly, it relies on the `Blockchain` event ordering.

> depends on the order of sends in core (?)

Yes, but that is a part of `Blockchain` interface guarantees. Maybe it's not yet written in the documentation, but that still is the case, even though there are several channels there they're used in a very particular order.

> and/or timing assumptions (?)

I don't think so.

We need to provide some consistent event ordering for subscribers, both at RPC and `Blockchain` levels. And I think we do provide it now. Now `Blockchain` has a different model for event distribution as it has four subscription channels, but that's intentional, this internal mechanism has a bit different usage model, multichannel system allows for more type safety and simplifies the most common use case (subscribing for blocks). On the other (RPC) side we may also use some additional event sources in the future (mempool being the most obvious one) and simple channel to event type mapping could also be useful there.

Reply:
> Maybe it's not yet written in the documentation

Actually it is already.